README.rst
=================================
txTemplate
=================================

txTemplate provides adapters for a few popular template engines
to make them easily callable and usable within Twisted Web.

txTemplate uses zope.interface to provide a consistent
loader and template interface for every template engine.

------------------------------------------
Supported Template Engines
------------------------------------------

 - ClearSilver
 - Genshi
 - Jinja2

------------------------------------------
Getting Started
------------------------------------------

All you really need to do to use txTemplate in twisted is:
 - import it
 - create a loader pointed at your template directory
 - load a template with said loader
 - call template.render, attach callbacks to the rendered deferred (if desired) and return the deferred from and of your twisted.web.resource.Resource render_* methods.

Here's a code sample of a Twisted Web resource that uses txTemplate
to render a lovely Genshi template to say Hello World in HTML::

    import os

    from twisted.application import internet
    from twisted.internet import reactor
    from twisted.web import resource
    from twisted.web import server

    import txtemplate


    TEMPLATE_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                "templates")


    class HelloWorld(resource.Resource):
        def __init__(self):
            resource.Resource.__init__(self)
            self.loader = txtemplate.GenshiTemplateLoader(TEMPLATE_DIR)

        def getChild(self, name, request):
            return self

        def render_GET(self, request):
            template_name = "hello.xhtml"
            template = self.loader.load(template_name)
            context = {"greeting": "Hello",
                    "greetee": "World"}

            def cb(content):
                request.write(content)
                request.setResponseCode(200)
                request.finish()

            d = template.render(**context)
            d.addCallback(cb)
            return server.NOT_DONE_YET


    site = server.Site(HelloWorld())
    reactor.listenTCP(8888, site)
    reactor.run()


This example is included in `txtemplate/examples/txweb.py` and the
`root.xhtml` template is in `txtemplate/examples/templates/root.xhtml`.





docs/changelog.txt
.. currentmodule:: mock


CHANGELOG
=========

2012/10/06 Version 1.0.2
--------------------------------

* Adding support for multiple template directories to the Jinja2 loader.
* Switched StringIO to cStringIO for performance and unicode support.

2012/06/17 Version 1.0.1
--------------------------------

* Performance improvements
 - Adding passthrough for jinja2 environment options so we can set auto_reload=False
 - updating tuning parameters CALL_DELAY and POPULATE_N_STEPS.
* Removed distribute_setup.py and requiring distribute/setuptools already be installed.

2012/02/29 Version 1.0.0
-------------------------

Initial release.


test/test_clearsilver.py
from twisted.trial import unittest
from twisted.internet import base
base.DelayedCall.debug = True

from txtemplate.templates import clearsilver

class TestClearsilver(unittest.TestCase):
    if clearsilver.CS is None:
        skip = "Skipping Clearsilver tests because Clearsilver is not installed."

    def setUp(self):
        self.hdf = clearsilver.HDF()

    def test_hdf_get_set_Value(self):
        page = "subcontract listing"
        self.hdf.setValue("page", page)
        self.assertEqual(page, self.hdf.getValue("page",""))
        
    


txtemplate/test/test_adapters.py
"""

Examples of using the twisted component registry to
lookup the appropriate template loader.

"""


from twisted.trial import unittest

from txtemplate import clearsilver
from txtemplate.itemplate import ITemplate
from txtemplate import templates


class TestIlookupClearsilver(unittest.TestCase):
    if clearsilver.CS is None:
        skip = "Skipping becasue clearsilver templates are not installed"

    def test_clearsilver(self):
        it = ITemplate(clearsilver.Template("foo"))
        self.assertTrue(it is not None)
        self.assertTrue(ITemplate.providedBy(it))


class TestILookupGenshi(unittest.TestCase):
    """Given a template I should be able to
    get an ITemplate adapter for that template that
    supports the necessary interface.

    """
    if templates.genshitemplate is None:
        skip = "Skipping because genshi templates are not installed"

    def test_genshitemplate(self):
        it = ITemplate(templates.genshitemplate.MarkupTemplate("<html></html>"))
        self.assertTrue(it is not None)
        self.assertTrue(ITemplate.providedBy(it))


class TestILookupJinja2(unittest.TestCase):
    """Given a template I should be able to
    get an ITemplate adapter for that template that
    supports the necessary interface.

    """
    if templates.jinja2 is None:
        skip = "Skipping because jinja2 templates are not installed"

    def test_jinja2template(self):
        it = ITemplate(templates.jinja2.Template("<html></html>"))
        self.assertTrue(it is not None)
        self.assertTrue(ITemplate.providedBy(it))





txtemplate/test/test_templates.py
from twisted.internet import defer, reactor
from twisted.trial import unittest
from twisted.internet import base
from twisted.python import failure, filepath
base.DelayedCall.debug = True

from txtemplate import templates
from txtemplate import clearsilver

TEMPLATE_DIR = filepath.FilePath(__file__
                                 ).parent().parent().child("templates").path


class TestClearsilverTemplates(unittest.TestCase):
    if clearsilver.CS is None:
        skip = "Skipping Clearsilver tests because Clearsilver is not installed."

    def setUp(self):
        self.c = {"request":"foo",
                  "segments":["hello","world"],
                  "message":"error message",
                  }

        self.loader = templates.ClearsilverTemplateLoader(
            TEMPLATE_DIR
            )

    class TestClass(object):
        someProperty = "Some Property Value"
        def __str__(self):
            return self.someProperty

    def test_loader_failure(self):
        self.assertRaises(templates.TemplateException, self.loader.load, "filethatdoesnotexist.cst")

    def test_loader_success(self):
        template = self.loader.load("test.cst")
        self.assertTrue(isinstance(template, templates.ClearsilverTemplateAdapter))

    def test_render_blocking(self):
        template = self.loader.load("test.cst")
        xhtml = template.render_blocking(**self.c)
        self.assertTrue(xhtml.find("foo") !=-1)

    def test_render_blocking_list(self):
        template = self.loader.load("test.cst")
        xhtml = template.render_blocking(**self.c)
        self.assertTrue(xhtml.find("<div>hello</div>") !=-1)

    def test_render_blocking_list_objects(self):
        template = self.loader.load("test.cst")
        self.c["segments"][0] = self.TestClass()
        xhtml = template.render_blocking(**self.c)
        self.assertTrue(xhtml.find("<div>Some Property Value</div>") !=-1)

    def test_render_blocking_list_of_lists(self):
        template = self.loader.load("test.cst")
        self.c["segments"][0] = ["h","e","l","l","o"]
        xhtml = template.render_blocking(**self.c)
        self.assertTrue(xhtml.find("<p>h</p>") !=-1)

    def test_render_blocking_dict(self):
        template = self.loader.load("test.cst")
        self.c['dict'] = {"test":"1 key",
                          "test_list":[1,2,3],
                          "test_dict":{"nested":"2 key",
                                       "nested_list":[4,5,6],
                                       },
                          }
        xhtml = template.render_blocking(**self.c)
        self.assertTrue(xhtml.find("<p>1 key</p>") !=-1)
        self.assertTrue(xhtml.find("<p>2 key</p>") !=-1)

    def test_render_blocking_hdf(self):
        template = self.loader.load("test.cst")
        hdf = clearsilver.hdfFromKwargs(**self.c)
        xhtml = template.render_blocking(hdf=hdf)
        self.assertTrue(xhtml.find("foo") !=-1)

    def _success(self, result):
        self.assertTrue(result.find("</html>")!=-1)

    def test_render_success(self):
        template = self.loader.load("test.cst")
        d = template.render(**self.c)
        d.addCallback(self._success)


class TestGenshiTemplates(unittest.TestCase):
    if templates.genshitemplate is None:
        skip = "Skipping Genshi tests because genshi is not installed."

    def setUp(self):
        self.c = {"request":"foo",
                  "segments":["hello","world"],
                  "message":"error message"
                  }
        self.loader = templates.GenshiTemplateLoader(TEMPLATE_DIR)

    def test_loader_failure(self):
        self.assertRaises(templates.TemplateException, self.loader.load, "filethatdoesnotexist.xhtml")

    def test_loader_success(self):
        template = self.loader.load("error.xhtml")
        self.assertTrue(isinstance(template, templates.GenshiTemplateAdapter))

    def test_render_blocking(self):
        template = self.loader.load("error.xhtml")
        xhtml = template.render_blocking(**self.c)
        self.assertTrue(xhtml.find("foo") !=-1)

    def test_failed(self):
        template = self.loader.load("error.xhtml")
        #self.assertRaises(templates.TemplateException, template._failed, failure.Failure(ValueError))
        f = failure.Failure(ValueError)
        errorMessage = "Failed to generate template"
        self.assertTrue(errorMessage in template._failed(f))

    def test_rendered(self):
        template = self.loader.load("error.xhtml")
        template._buffer.write("Some string o' crap!")
        result = template._rendered(None)
        self.assertEqual(result, "Some string o' crap!")

    def _success(self, result):
        self.assertTrue(result.find("</html>")!=-1)

    def test_populateBuffer(self):
        template = self.loader.load("error.xhtml")
        template._deferred = defer.Deferred()
        template._deferred.addCallbacks(template._rendered, template._failed)
        template._deferred.addCallback(self._success)
        template._stream = template.template.generate(**self.c)
        s = template._stream.serialize()
        reactor.callLater(0.01, template._populateBuffer, s, templates.POPULATE_N_STEPS)
        return template._deferred

    def test_render(self):
        template = self.loader.load("error.xhtml")
        d = template.render(**self.c)
        d.addCallback(self._success)
        return d


class TestJinja2Templates(unittest.TestCase):
    if templates.jinja2 is None:
        skip = "Skipping Jinja2 tests because jinja2 is not installed."

    def setUp(self):
        self.c = {"request":"foo",
                  "segments":["hello","world"],
                  "message":"error message"
                  }
        self.loader = templates.Jinja2TemplateLoader(TEMPLATE_DIR)

    def test_loader_failure(self):
        self.assertRaises(templates.TemplateException, self.loader.load, "filethatdoesnotexist.jinja2")

    def test_loader_success(self):
        template = self.loader.load("test.jinja2")
        self.assertTrue(isinstance(template, templates.Jinja2TemplateAdapter))

    def test_render_blocking(self):
        template = self.loader.load("test.jinja2")
        html = template.render_blocking(**self.c)
        self.assertTrue(html.find("foo") !=-1)

    def _success(self, result):
        self.assertTrue(result.find("foo")!=-1)
    def test_render(self):
        template = self.loader.load("test.jinja2")
        d = template.render(**self.c)
        d.addCallback(self._success)
        return d


class TestJinja2MultiplePaths(unittest.TestCase):
    if templates.jinja2 is None:
        skip = "Skipping Jinja2 tests because jinja2 is not installed."

    def test_no_dir(self):
        loader = templates.Jinja2TemplateLoader(None)
        self.assertEqual(loader.paths, ['.'])

    def test_no_dir_and_options(self):
        loader = templates.Jinja2TemplateLoader(None, auto_reload=True)
        self.assertEqual(loader.paths, ['.'])

    def test_one_dir(self):
        loader = templates.Jinja2TemplateLoader([TEMPLATE_DIR])
        self.assertEqual(loader.paths, [TEMPLATE_DIR])

    def test_multiple_dirs(self):
        loader = templates.Jinja2TemplateLoader([TEMPLATE_DIR, TEMPLATE_DIR])
        self.assertEqual(loader.paths, [TEMPLATE_DIR] * 2)

    def test_multiple_dirs_and_options(self):
        loader = templates.Jinja2TemplateLoader([TEMPLATE_DIR, TEMPLATE_DIR], auto_reload=True)
        self.assertEqual(loader.paths, [TEMPLATE_DIR] * 2)


class TestStringTemplates(unittest.TestCase):
    def setUp(self):
        self.c = {"request":"foo",
                  "segments":["hello","world"],
                  "message":"error message",
                  }

        self.loader = templates.StringTemplateLoader(
            TEMPLATE_DIR
            )

    class TestClass(object):
        someProperty = "Some Property Value"
        def __str__(self):
            return self.someProperty

    def test_loader_failure(self):
        self.assertRaises(templates.TemplateException, self.loader.load, "filethatdoesnotexist.cst")

    def test_loader_success(self):
        template = self.loader.load("test.txt")
        self.assertTrue(isinstance(template, templates.StringTemplateAdapter))

    def test_render_blocking(self):
        template = self.loader.load("test.txt")
        xhtml = template.render_blocking(**self.c)
        self.assertTrue(xhtml.find("foo") !=-1)

    def _success(self, result):
        self.assertTrue(result.find("</html>")!=-1)

    def test_render_success(self):
        template = self.loader.load("test.txt")
        d = template.render(**self.c)
        d.addCallback(self._success)


clearsilver.py
#-*- test-case-name: txtemplate.test.test_clearsilver -*-
"""
Clearsilver library integration point

Handles imports of clearsilver modules and
provides helper routines to handle common
operations on hdf objects.
"""

try:
    import neo_cgi # neo_cgi must be imported before neo_cs or neo_util
    import neo_cs
    from neo_cs import CS
    import neo_util
    from neo_util import HDF
except ImportError, e:
    neo_cgi, neo_cs, neo_util, CS, HDF = None, None, None, None, None


class Template(object):
    def __init__(self, template_path):
        self.template_path = template_path



def hdfFromKwargs(hdf=None, **kwargs):
    """If given an instance that has toHDF() method that method is invoked to get that object's HDF representation"""
    if not hdf:
        hdf = HDF()
    for key, value in kwargs.iteritems():
        if isinstance(value, dict):
            #print "dict:",value
            for k,v in value.iteritems():
                dkey = "%s.%s"%(key,k)
                #print "k,v,dkey:",k,v,dkey
                args = {dkey:v}
                hdfFromKwargs(hdf=hdf, **args)
        elif isinstance(value, (list, tuple)):
            #print "list:",value
            for i, item in enumerate(value):
                ikey = "%s.%s"%(key,i)
                #print "i,item:",i,item, ikey
                if isinstance(item, (list, tuple)):
                    args = {ikey:item}
                    hdfFromKwargs(hdf=hdf, **args)
                elif isinstance(item, dict):
                    args = {ikey:item}
                    hdfFromKwargs(hdf=hdf, **args)
                elif getattr(item, "HDF_ATTRIBUTES", False):
                    attrs = {}
                    for attr in item.HDF_ATTRIBUTES:
                        attrs[attr] = getattr(item, attr, "")
                    hdfFromKwargs(hdf=hdf, **{ikey:attrs})
                else:
                    hdf.setValue(ikey, str(item))
        elif getattr(value, "HDF_ATTRIBUTES", False):
            attrs = {}
            for attr in value.HDF_ATTRIBUTES:
                attrs[attr] = getattr(value, attr, "")
            hdfFromKwargs(hdf=hdf, **{key:attrs})
        else:
            hdf.setValue(key, str(value))
    #print "HDF:",hdf.dump()
    return hdf



txtemplate/itemplate.py
"""itemplate.py

Interfaces for the txTemplate
"""

from zope import interface


class ITemplateLoader(interface.Interface):
    """
    Template loaders simply take a template name and return an instance of a
    template object.

    Arguably this would be better as a simple IFile adapter.
    """

    def load(name):
        """
        Load the template from a unique identifier, usually the file name
        without the path.  The loader itself knows the path to the template.
        """


class ITemplate(interface.Interface):
    """
    A document, used as a pattern for generating content via the
    insertion of data into predefined locations within the document.
    """

    def render(**kwargs):
        """
        Return an instance of twisted.internet.defer.Deferred, which will fire
        with the rendered string, or, in the case of an error, an instance of
        txtemplate.templates.TemplateException.

        Examples:
            def failureFunction(reason):
                print reason

            def successFunction(renderedString):
                print renderedString

            data = {'key1': value1, 'key2': value2}
            d1 = template.render(**data)
            d1.addCallbacks(successFunction, failureFunction)

            d2 = template.render(key1=value1, key2=value2)
            d2.addCallbacks(successFunction, failureFunction)
        """

    def render_blocking(**kwargs):
        """
        Render my content as a string, using the key-value pairs provided as
        arguments to this method.

        Examples:
            data = {'key1': value1, 'key2': value2}
            s1 = template.render_blocking(**data)

            s2 = template.render_blocking(key1=value1, key2=value2)
        """





txtemplate/templates.py
"""
Templates support for Twisted projects
"""

import StringIO
import os
import string

try:
    from genshi import template as genshitemplate
except ImportError, e:
    genshitemplate = None

try:
    import jinja2
except ImportError, e:
    jinja2 = None

from twisted.internet import defer, reactor
from twisted.python import components
from zope import interface

from txtemplate import clearsilver
from txtemplate import itemplate

CALL_DELAY = 0.00001
POPULATE_N_STEPS = 10000


class TemplateException(Exception):
    """
    Umbrella exception for all template processing errors

    Anyone interested in catching all template related
    exceptions should simply catch this.
    """
    pass


def registerIfNotRegistered(adapter, from_, to):
    if not components.getAdapterFactory(from_, to, None):
        components.registerAdapter(adapter, from_, to)


class ClearsilverTemplateAdapter(object):
    """
    Adapter that provides itemplate.ITemplate
    """

    interface.implements(itemplate.ITemplate)

    def __init__(self, template):
        self.template = template

    def render_blocking(self, hdf=None, **kwargs):
        hdf = clearsilver.hdfFromKwargs(hdf=hdf, **kwargs)
        cs = clearsilver.CS(hdf)
        cs.parseFile(self.template.template_path)
        result = cs.render()
        return result

    def _prepareHDF(self):
        try:
            hdf = clearsilver.hdfFromKwargs(hdf=self.hdf,
                                        **self.kwargs)
        except Exception, e:
            self._deferred.errback(e)
        else:
            self._deferred.callback(hdf)

    def _failedHDF(self, error):
        print error

    def _rendered(self, hdf):
        cs = clearsilver.CS(hdf)
        cs.parseFile(self.template.template_path)
        result = cs.render()
        return result

    def _failed(self, reason):
        raise TemplateException(
            "Failed to generate template %s because %s"%(
            self.template, reason)
            )

    def render(self, hdf=None, **kwargs):
        self.hdf = hdf
        self.kwargs = kwargs
        self._deferred = defer.Deferred()
        self._deferred.addCallbacks(self._rendered, self._failed)
        reactor.callLater(CALL_DELAY, self._prepareHDF)
        return self._deferred


if clearsilver.CS is not None:
    registerIfNotRegistered(
        ClearsilverTemplateAdapter,
        clearsilver.Template,
        itemplate.ITemplate
    )


class ClearsilverTemplateLoader(object):
    interface.implements(itemplate.ITemplateLoader)

    def __init__(self, path=None):
        self.path = path
        if not self.path:
            self.path = os.curdir

    def load(self, name):
        template_path = os.path.join(self.path, name)
        if not os.path.exists(template_path):
            raise TemplateException("Template %s not found! (expected: %s)"%(name, template_path))
        else:
            template = clearsilver.Template(template_path
                                            )
            return itemplate.ITemplate(template)


if clearsilver.CS is not None:
    registerIfNotRegistered(
        ClearsilverTemplateLoader,
        ClearsilverTemplateLoader,
        itemplate.ITemplateLoader
    )


class GenshiTemplateAdapter(object):
    """
    Adapter that provides itemplate.ITemplate for Genshi Templates
    """

    interface.implements(itemplate.ITemplate)

    def __init__(self, template):
        self._buffer = StringIO.StringIO()
        self._stream = None
        self.template = template
        self.delayedCall = None

    def _populateBuffer(self, stream, n):
        """
        Iterator that returns N steps of
        the genshi stream.

        Found that performance really sucks for
        n = 1 (0.5 requests/second for the root resources
        versus 80 requests/second for a blocking algorithm).

        Hopefully increasing the number of steps per timeslice will
        significantly improve performance.
        """
        try:
            for x in xrange(n):
                output = stream.next()
                self._buffer.write(output)
        except StopIteration, e:
            self._deferred.callback(None)
        except Exception, e:
            self._deferred.errback(e)
        else:
            self.delayedCall = reactor.callLater(CALL_DELAY, self._populateBuffer, stream, n)

    def _failed(self, reason):
        if self.delayedCall and self.delayedCall.active():
            self.delayedCall.cancel()
        return "Failed to generate template %s because %s"%(self.template, reason)

    def _rendered(self, ignore):
        result = self._buffer.getvalue()
        self._buffer.close()
        self._buffer = None
        if self.delayedCall and self.delayedCall.active():
            self.delayedCall.cancel()
        return result.encode('UTF-8')

    def render(self, **kwargs):
        self._stream = self.template.generate(**kwargs)
        self._deferred = defer.Deferred()
        self._deferred.addCallbacks(self._rendered, self._failed)
        s = self._stream.serialize()
        self.delayedCall = reactor.callLater(CALL_DELAY, self._populateBuffer, s, POPULATE_N_STEPS)
        return self._deferred

    def render_blocking(self, **kwargs):
        tmp = self.template.generate(**kwargs)
        return tmp.render()


if genshitemplate is not None:
    registerIfNotRegistered(
        GenshiTemplateAdapter,
        genshitemplate.MarkupTemplate,
        itemplate.ITemplate
    )


class GenshiTemplateLoader(object):
    interface.implements(itemplate.ITemplateLoader)

    def __init__(self, path=None):
        self.path = path
        if not self.path:
            self.path = os.curdir

        self.loader = genshitemplate.TemplateLoader(
            os.path.join(os.path.abspath(self.path)), auto_reload=True
        )

    def load(self, name):
        try:
            template = self.loader.load(name, encoding="UTF-8")
        except genshitemplate.loader.TemplateNotFound:
            raise TemplateException("Template %s not found!"%(name))
        else:
            return itemplate.ITemplate(template)


if genshitemplate is not None:
    registerIfNotRegistered(
        GenshiTemplateLoader,
        genshitemplate.TemplateLoader,
        itemplate.ITemplateLoader
    )


class Jinja2TemplateAdapter(GenshiTemplateAdapter):
    interface.implements(itemplate.ITemplate)

    def render(self, **kwargs):
        iterator = self.template.generate(**kwargs)
        self._deferred = defer.Deferred()
        self._deferred.addCallbacks(self._rendered, self._failed)
        self.delayedCall = reactor.callLater(
            CALL_DELAY, self._populateBuffer,
            iterator, POPULATE_N_STEPS)
        return self._deferred

    def render_blocking(self, **kwargs):
        return self.template.render(**kwargs)


if jinja2 is not None:
    registerIfNotRegistered(
        Jinja2TemplateAdapter,
        jinja2.Template,
        itemplate.ITemplate
    )


class Jinja2TemplateLoader(object):
    interface.implements(itemplate.ITemplateLoader)

    def __init__(self, paths, **options):
        """

        args:
         - `paths`: a path or iterable of paths to search for templates
           NOTE::
             txtemplate ensures they are all absolute paths before
             passing them to the underlying template engine.

        Supported options:
         - `auto_reload`: providing `auto_reload=True` will cause Jinja2
           to check the filesystem for changes every time it tries to render
           a template to ensure it has the most up to date template.

        """
        if not paths:
            self.paths = [os.curdir]
        elif isinstance(paths, basestring):
            self.paths = [paths]
        else:
            self.paths = paths
        self.options = {"auto_reload": False}
        self.options.update(options)

        self.loader = jinja2.FileSystemLoader(
            [os.path.abspath(p) for p in self.paths],
            encoding="utf-8"
        )
        self.environment = jinja2.Environment(loader=self.loader, **self.options)

    def load(self, name):
        try:
            template = self.environment.get_template(name)
        except jinja2.exceptions.TemplateNotFound:
            raise TemplateException("Template %s not found!"%(name))
        else:
            return itemplate.ITemplate(template)


if jinja2 is not None:
    registerIfNotRegistered(
        Jinja2TemplateLoader,
        jinja2.FileSystemLoader,
        itemplate.ITemplateLoader
    )


class StringTemplateAdapter(object):
    """Adapter that provides itemplate.ITemplate for
    Python's standard string.Template.  This provides
    a basic level of functionality without requiring a
    real templating library.
    """
    interface.implements(itemplate.ITemplate)
    def __init__(self, template):
        self.template = template

    def render_blocking(self, **kwargs):
        rendered = self.template.substitute(kwargs)
        return rendered

    def _render(self, kwargs):
        try:
            rendered = self.template.substitute(kwargs)
        except Exception, e:
            self._deferred.errback(e)
        else:
            self._deferred.callback(rendered)

    def _rendered(self, rendered):
        return rendered

    def _failed(self, reason):
        raise TemplateException(
            "Failed to generate template %s because %s"%(
            self.template, reason)
            )

    def render(self, **kwargs):
        self._deferred = defer.Deferred()
        self._deferred.addCallbacks(self._rendered, self._failed)
        reactor.callLater(CALL_DELAY, self._render, kwargs)
        return self._deferred


registerIfNotRegistered(
        StringTemplateAdapter,
        string.Template,
        itemplate.ITemplate
)


class StringTemplateLoader(object):
    interface.implements(itemplate.ITemplateLoader)

    def __init__(self, path=None):
        self.path = path
        if not self.path:
            self.path = os.curdir

    def load(self, name):
        template_path = os.path.join(self.path, name)
        if not os.path.exists(template_path):
            raise TemplateException("Template %s not found! (expected: %s)"%(name, template_path))
        else:
            templateContent = open(template_path, "r").read()
            template = string.Template(templateContent)
            return itemplate.ITemplate(template)

registerIfNotRegistered(
        StringTemplateLoader,
        StringTemplateLoader,
        itemplate.ITemplateLoader
)


txtemplate/examples/txweb.py
"""
Twisted Web example using txTemplate.
"""

import os

from twisted.application import internet
from twisted.internet import reactor
from twisted.web import resource
from twisted.web import server

import txtemplate


TEMPLATE_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                            "templates")


class HelloWorld(resource.Resource):
    def __init__(self):
        resource.Resource.__init__(self)
        self.loader = txtemplate.GenshiTemplateLoader(TEMPLATE_DIR)

    def getChild(self, name, request):
        return self

    def render_GET(self, request):
        template_name = "hello.xhtml"
        template = self.loader.load(template_name)
        context = {"greeting": "Hello",
                "greetee": "World"}

        def cb(content):
            request.write(content)
            request.setResponseCode(200)
            request.finish()

        d = template.render(**context)
        d.addCallback(cb)
        return server.NOT_DONE_YET


site = server.Site(HelloWorld())
reactor.listenTCP(8888, site)
reactor.run()


